File: packages/contracts/src/calendar/trading-calendar.ts
/**
 * A-share trading calendar.
 *
 * Normalizes trading days, sessions, holidays and temporary closures into one
 * queryable calendar. Times are local clock HH:MM in Asia/Shanghai; sessions
 * are half-open [start, end). A session that crosses midnight (start > end,
 * e.g. a night session 21:00 -> 02:30) belongs to the trading day on which it
 * starts. The default templates and seed holidays come from
 * packages/contracts/calendar/ — the single source of truth shared with the
 * Python port in market-core.
 */

import sessionTemplates from "../../calendar/session-templates.json";
import holidaysJson from "../../calendar/holidays.json";

export type ExchangeId = "SH" | "SZ" | "BJ";
export type SessionKind = "CONTINUOUS" | "OPEN_AUCTION" | "CLOSE_AUCTION" | "BREAK" | "NIGHT";
export type HolidayReason = "PUBLIC_HOLIDAY" | "TEMPORARY_CLOSURE" | "SPECIAL";
export type NonTradingReason = "WEEKEND" | HolidayReason;

export interface TradingSession {
  /** Stable machine id, e.g. open-auction / morning / lunch-break / afternoon. */
  sessionId: string;
  kind: SessionKind;
  /** Local clock HH:MM (Asia/Shanghai), inclusive start. */
  start: string;
  /** Local clock HH:MM (Asia/Shanghai), exclusive end. */
  end: string;
  exchange: ExchangeId;
  /** True when the session ends after midnight (start > end). */
  crossesMidnight?: boolean;
}

export interface TradingDay {
  /** ISO 8601 calendar date in Asia/Shanghai. */
  date: string;
  exchange: ExchangeId;
  isTradingDay: boolean;
  reason?: NonTradingReason;
  note?: string;
}

export interface Holiday {
  date: string;
  exchange: ExchangeId;
  reason: HolidayReason;
  note?: string;
}

export interface SessionAtResult {
  session: TradingSession;
  /** The trading day the session belongs to (may be the previous calendar day). */
  date: string;
}

export interface TradingCalendar {
  /** Sessions that apply to a given trading day; empty when not a trading day. */
  sessionsFor(date: string, exchange: ExchangeId): TradingSession[];
  /** Whether the date is a trading day for the exchange (weekends/holidays are never trading days). */
  isTradingDay(date: string, exchange: ExchangeId): boolean;
  /** The per-date status record for an exchange. */
  tradingDay(date: string, exchange: ExchangeId): TradingDay;
  /** The session containing a given instant, if any. */
  sessionAt(instant: Date, exchange: ExchangeId): SessionAtResult | null;
  /** The first trading day strictly after the given date. */
  nextTradingDay(date: string, exchange: ExchangeId): string;
}

const DEFAULT_TEMPLATES = sessionTemplates as unknown as {
  timezone: string;
  exchanges: Record<ExchangeId, { sessions: TradingSession[] }>;
};
const DEFAULT_HOLIDAYS = holidaysJson as unknown as { timezone: string; holidays: Holiday[] };

export const MARKET_TIMEZONE = "Asia/Shanghai";

const LOCAL_FORMATTER = new Intl.DateTimeFormat("en-CA", {
  timeZone: MARKET_TIMEZONE,
  year: "numeric",
  month: "2-digit",
  day: "2-digit",
  hour: "2-digit",
  minute: "2-digit",
  second: "2-digit",
  hourCycle: "h23",
});

function toLocalParts(instant: Date): { date: string; time: string } {
  const parts = LOCAL_FORMATTER.formatToParts(instant);
  const get = (type: Intl.DateTimeFormatPartTypes): string =>
    parts.find((p) => p.type === type)?.value ?? "";
  return {
    date: `${get("year")}-${get("month")}-${get("day")}`,
    time: `${get("hour")}:${get("minute")}:${get("second")}`,
  };
}

/** "HH:MM" -> minutes since local midnight. */
function toMinutes(clock: string): number {
  const [h, m] = clock.split(":");
  return Number(h) * 60 + Number(m);
}

function addDays(date: string, days: number): string {
  const d = new Date(`${date}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

function weekday(date: string): number {
  // 0 = Sunday, 6 = Saturday.
  return new Date(`${date}T00:00:00Z`).getUTCDay();
}

function isWeekend(date: string): boolean {
  const w = weekday(date);
  return w === 0 || w === 6;
}

/**
 * Whether a local clock time falls inside a session. Cross-midnight sessions
 * cover [start, 24:00) ∪ [00:00, end) of the same (start) trading day.
 */
function inSession(session: TradingSession, minutes: number): boolean {
  const start = toMinutes(session.start);
  const end = toMinutes(session.end);
  if (session.crossesMidnight || start > end) {
    return minutes >= start || minutes < end;
  }
  return minutes >= start && minutes < end;
}

/**
 * Default in-memory calendar built from the shared templates and seed holidays.
 * Real provider-synced calendars (P1-N07) will implement the same interface.
 */
export class StaticTradingCalendar implements TradingCalendar {
  private readonly templates: Record<string, TradingSession[]>;
  private readonly holidaysByKey: Map<string, Holiday>;

  constructor(
    templates: Record<string, TradingSession[]> = Object.fromEntries(
      Object.entries(DEFAULT_TEMPLATES.exchanges).map(([exchange, t]) => [exchange, t.sessions]),
    ),
    holidays: Holiday[] = DEFAULT_HOLIDAYS.holidays,
  ) {
    this.templates = templates;
    this.holidaysByKey = new Map(holidays.map((h) => [`${h.date}:${h.exchange}`, h]));
  }

  sessionsFor(date: string, exchange: string): TradingSession[] {
    return this.isTradingDay(date, exchange) ? (this.templates[exchange] ?? []) : [];
  }

  isTradingDay(date: string, exchange: string): boolean {
    return !isWeekend(date) && !this.holidaysByKey.has(`${date}:${exchange}`);
  }

  tradingDay(date: string, exchange: string): TradingDay {
    const holiday = this.holidaysByKey.get(`${date}:${exchange}`);
    const exchangeId = exchange as ExchangeId;
    if (holiday) {
      return { date, exchange: exchangeId, isTradingDay: false, reason: holiday.reason, note: holiday.note };
    }
    if (isWeekend(date)) {
      return { date, exchange: exchangeId, isTradingDay: false, reason: "WEEKEND" };
    }
    return { date, exchange: exchangeId, isTradingDay: true };
  }

  sessionAt(instant: Date, exchange: string): SessionAtResult | null {
    const { date, time } = toLocalParts(instant);
    const minutes = toMinutes(time.slice(0, 5));

    const sessions = this.templates[exchange] ?? [];
    // A cross-midnight session belongs to its start trading day: the evening
    // part [start, 24:00) matches the start day, while the early morning part
    // [00:00, end) is matched by the *previous* day's session below.
    if (this.isTradingDay(date, exchange)) {
      for (const session of sessions) {
        const crossesMidnight = session.crossesMidnight || toMinutes(session.start) > toMinutes(session.end);
        if (crossesMidnight) {
          if (minutes >= toMinutes(session.start)) {
            return { session, date };
          }
        } else if (inSession(session, minutes)) {
          return { session, date };
        }
      }
    }

    const yesterday = addDays(date, -1);
    if (this.isTradingDay(yesterday, exchange)) {
      for (const session of sessions) {
        const crossesMidnight = session.crossesMidnight || toMinutes(session.start) > toMinutes(session.end);
        if (crossesMidnight && minutes < toMinutes(session.end)) {
          return { session, date: yesterday };
        }
      }
    }
    return null;
  }

  nextTradingDay(date: string, exchange: string): string {
    let candidate = addDays(date, 1);
    for (let guard = 0; guard < 366; guard += 1) {
      if (this.isTradingDay(candidate, exchange)) {
        return candidate;
      }
      candidate = addDays(candidate, 1);
    }
    throw new Error(`No trading day found within 365 days after ${date}`);
  }
}

/** The default A-share calendar (SH/SZ/BJ session templates + seed holidays). */
export const DEFAULT_CALENDAR: TradingCalendar = new StaticTradingCalendar();

File: python/packages/market-core/src/market_core/calendar/trading_calendar.py
"""A-share trading calendar.

Python port of the TypeScript calendar in packages/contracts/src/calendar.
Both read the same session templates and seed holidays from
packages/contracts/calendar/ — the single source of truth.

Times are local clock HH:MM in Asia/Shanghai; sessions are half-open
[start, end). A session that crosses midnight (start > end, e.g. a night
session 21:00 -> 02:30) belongs to the trading day on which it starts.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import date, datetime, timedelta
from pathlib import Path
from typing import Literal, Protocol, cast
from zoneinfo import ZoneInfo

ExchangeId = Literal["SH", "SZ", "BJ"]
SessionKind = Literal["CONTINUOUS", "OPEN_AUCTION", "CLOSE_AUCTION", "BREAK", "NIGHT"]
HolidayReason = Literal["PUBLIC_HOLIDAY", "TEMPORARY_CLOSURE", "SPECIAL"]
NonTradingReason = Literal["WEEKEND", "PUBLIC_HOLIDAY", "TEMPORARY_CLOSURE", "SPECIAL"]

MARKET_TIMEZONE = "Asia/Shanghai"
_TZ = ZoneInfo(MARKET_TIMEZONE)

_CONTRACTS_DIR = Path(__file__).resolve().parents[6] / "packages" / "contracts" / "calendar"


@dataclass(frozen=True)
class TradingSession:
    """A named time interval within a trading day."""

    session_id: str
    kind: SessionKind
    start: str
    end: str
    exchange: ExchangeId
    crosses_midnight: bool = False


@dataclass(frozen=True)
class TradingDay:
    """Per-date status record for an exchange."""

    date: str
    exchange: ExchangeId
    is_trading_day: bool
    reason: NonTradingReason | None = None
    note: str | None = None


@dataclass(frozen=True)
class Holiday:
    """A scheduled exchange closure (public holiday or ad-hoc closure)."""

    date: str
    exchange: ExchangeId
    reason: HolidayReason
    note: str | None = None


@dataclass(frozen=True)
class SessionAtResult:
    """The session containing an instant and the trading day it belongs to."""

    session: TradingSession
    date: str


class TradingCalendar(Protocol):
    """Queryable calendar over trading days and sessions.

    Real provider-synced calendars (P1-N07) must implement the same interface.
    """

    def sessions_for(self, day: str, exchange: str) -> list[TradingSession]: ...

    def is_trading_day(self, day: str, exchange: str) -> bool: ...

    def trading_day(self, day: str, exchange: str) -> TradingDay: ...

    def session_at(self, instant: datetime, exchange: str) -> SessionAtResult | None: ...

    def next_trading_day(self, day: str, exchange: str) -> str: ...


def _to_session(raw: dict[str, object]) -> TradingSession:
    return TradingSession(
        session_id=str(raw["sessionId"]),
        kind=cast(SessionKind, raw["kind"]),
        start=str(raw["start"]),
        end=str(raw["end"]),
        exchange=cast(ExchangeId, raw["exchange"]),
        crosses_midnight=bool(raw.get("crossesMidnight", False)),
    )


def _to_holiday(raw: dict[str, object]) -> Holiday:
    return Holiday(
        date=str(raw["date"]),
        exchange=cast(ExchangeId, raw["exchange"]),
        reason=cast(HolidayReason, raw["reason"]),
        note=raw.get("note"),  # type: ignore[arg-type]
    )


def _load_rules() -> tuple[dict[str, list[TradingSession]], list[Holiday]]:
    with (_CONTRACTS_DIR / "session-templates.json").open(encoding="utf-8") as f:
        templates = json.load(f)
    with (_CONTRACTS_DIR / "holidays.json").open(encoding="utf-8") as f:
        holidays = json.load(f)
    exchange_templates = cast(dict[str, dict[str, object]], templates["exchanges"])
    sessions = {
        key: [
            _to_session(cast(dict[str, object], s)) for s in cast(list[object], value["sessions"])
        ]
        for key, value in exchange_templates.items()
    }
    holiday_models = [
        _to_holiday(cast(dict[str, object], h)) for h in cast(list[object], holidays["holidays"])
    ]
    return sessions, holiday_models


def _to_minutes(clock: str) -> int:
    hour, minute = clock.split(":")
    return int(hour) * 60 + int(minute)


def _add_days(day: str, days: int) -> str:
    return (date.fromisoformat(day) + timedelta(days=days)).isoformat()


def _is_weekend(day: str) -> bool:
    return date.fromisoformat(day).weekday() >= 5  # Sat=5, Sun=6


class StaticTradingCalendar:
    """Default in-memory calendar built from shared templates and seed holidays.

    ``templates`` maps an exchange key to its sessions and ``holidays`` is the
    closure list; both default to the shared rule tables in
    packages/contracts/calendar/.
    """

    def __init__(
        self,
        templates: dict[str, list[TradingSession]] | None = None,
        holidays: list[Holiday] | None = None,
    ) -> None:
        default_templates, default_holidays = _load_rules()
        self._templates = templates if templates is not None else default_templates
        effective_holidays = holidays if holidays is not None else default_holidays
        self._holidays_by_key: dict[str, Holiday] = {
            f"{h.date}:{h.exchange}": h for h in effective_holidays
        }

    def sessions_for(self, day: str, exchange: str) -> list[TradingSession]:
        return self._templates.get(exchange, []) if self.is_trading_day(day, exchange) else []

    def is_trading_day(self, day: str, exchange: str) -> bool:
        return not _is_weekend(day) and f"{day}:{exchange}" not in self._holidays_by_key

    def trading_day(self, day: str, exchange: str) -> TradingDay:
        holiday = self._holidays_by_key.get(f"{day}:{exchange}")
        if holiday is not None:
            return TradingDay(
                date=day,
                exchange=cast(ExchangeId, exchange),
                is_trading_day=False,
                reason=holiday.reason,
                note=holiday.note,
            )
        if _is_weekend(day):
            return TradingDay(
                date=day,
                exchange=cast(ExchangeId, exchange),
                is_trading_day=False,
                reason="WEEKEND",
            )
        return TradingDay(date=day, exchange=cast(ExchangeId, exchange), is_trading_day=True)

    def session_at(self, instant: datetime, exchange: str) -> SessionAtResult | None:
        local = instant.astimezone(_TZ)
        day = local.date().isoformat()
        minutes = local.hour * 60 + local.minute
        sessions = self._templates.get(exchange, [])

        # A cross-midnight session belongs to its start trading day: the
        # evening part [start, 24:00) matches the start day, while the early
        # morning part [00:00, end) is matched by the *previous* day below.
        if self.is_trading_day(day, exchange):
            for session in sessions:
                start = _to_minutes(session.start)
                end = _to_minutes(session.end)
                crosses = session.crosses_midnight or start > end
                if crosses:
                    if minutes >= start:
                        return SessionAtResult(session, day)
                elif start <= minutes < end:
                    return SessionAtResult(session, day)

        yesterday = _add_days(day, -1)
        if self.is_trading_day(yesterday, exchange):
            for session in sessions:
                start = _to_minutes(session.start)
                end = _to_minutes(session.end)
                crosses = session.crosses_midnight or start > end
                if crosses and minutes < end:
                    return SessionAtResult(session, yesterday)
        return None

    def next_trading_day(self, day: str, exchange: str) -> str:
        candidate = _add_days(day, 1)
        for _ in range(366):
            if self.is_trading_day(candidate, exchange):
                return candidate
            candidate = _add_days(candidate, 1)
        raise RuntimeError(f"No trading day found within 365 days after {day}")


# The default A-share calendar (SH/SZ/BJ session templates + seed holidays).
DEFAULT_CALENDAR: TradingCalendar = StaticTradingCalendar()
